// react-charts/lib/hooks/useRect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];

exports.__esModule = true;
exports["default"] = useRect;

var _react = _interopRequireDefault(require("react"));

function useRect(element, enabled) {
  var rerender = _react["default"].useReducer(function () {
    return {};
  }, [])[1];

  var rectRef = _react["default"].useRef({
    width: 0,
    height: 0,
    x: 0,
    y: 0,
    left: 0,
    top: 0,
    right: 0,
    bottom: 0
  });

  var measure = _react["default"].useCallback(function () {
    if (element) {
      rectRef.current = element.getBoundingClientRect();
    }
  }, [element]);

  if (!rectRef.current) {
    measure();
  }

  _react["default"].useEffect(function () {
    if (!element || !enabled) {
      return;
    }

    var cb = function cb() {
      measure();
      rerender();
    };

    document.addEventListener('scroll', cb, true);
    return function () {
      document.removeEventListener('scroll', cb, true);
    };
  }, [element, enabled, measure, rerender]);

  _react["default"].useEffect(function () {
    if (!element || !enabled) {
      return;
    }

    measure();
    rerender();
    var observer = new ResizeObserver(function () {
      measure();
      rerender();
    });
    observer.observe(element);
    return function () {
      observer.unobserve(element);
    };
  }, [element, enabled, measure, rerender]);

  return rectRef.current;
}
